api/etcd/list.go

```diff
@@ -4,200 +4,61 @@ import (
 	"context"
 	"encoding/base64"
 	"strconv"
-	"strings"
 
 	"github.com/coreos/etcd/clientv3"
-	"github.com/gogo/protobuf/proto"
-	"github.com/heetch/regula"
 	"github.com/heetch/regula/api"
-	pb "github.com/heetch/regula/api/etcd/proto"
 	"github.com/pkg/errors"
 )
 
-func computeLimit(l int) int {
-	if l <= 0 || l > 100 {
-		return 50 // TODO: make this one configurable
-	}
-	return l
-}
-
-// List returns the latest version of each ruleset under the given prefix.
-// If the prefix is empty, it returns rulesets from the beginning following the lexical order.
-// The listing can be customised using the ListOptions type.
-func (s *RulesetService) List(ctx context.Context, prefix string, opt *api.ListOptions) (*api.Rulesets, error) {
-	options := make([]clientv3.OpOption, 0, 3)
+// List returns a list of ruleset paths.
+// The listing is paginated and can be customised using the ListOptions type.
+func (s *RulesetService) List(ctx context.Context, opt api.ListOptions) (*api.Rulesets, error) {
+	var opts []clientv3.OpOption
 
 	var key string
 
-	limit := computeLimit(opt.Limit)
+	// only fetch keys
+	opts = append(opts, clientv3.WithKeysOnly())
 
-	if opt.ContinueToken != "" {
-		lastPath, err := base64.URLEncoding.DecodeString(opt.ContinueToken)
+	// if a cursor is specified, decode the key from it and start the request from that key
+	if opt.Cursor != "" {
+		lastPath, err := base64.URLEncoding.DecodeString(opt.Cursor)
 		if err != nil {
-			return nil, api.ErrInvalidContinueToken
+			return nil, api.ErrInvalidCursor
 		}
 
-		key = string(lastPath)
+		key = s.signaturesPath(string(lastPath))
 
-		var rangeEnd string
-		if opt.AllVersions {
-			rangeEnd = clientv3.GetPrefixRangeEnd(s.rulesPath(prefix, ""))
-		} else {
-			rangeEnd = clientv3.GetPrefixRangeEnd(s.latestVersionPath(prefix))
-		}
-		options = append(options, clientv3.WithRange(rangeEnd))
+		opts = append(opts, clientv3.WithRange(clientv3.GetPrefixRangeEnd(s.signaturesPath(""))))
 	} else {
-		key = prefix
-		options = append(options, clientv3.WithPrefix())
+		key = s.signaturesPath("")
+		opts = append(opts, clientv3.WithPrefix())
 	}
 
-	options = append(options, clientv3.WithLimit(int64(limit)))
-
-	switch {
-	case opt.PathsOnly:
-		return s.listPathsOnly(ctx, key, prefix, limit, options)
-	case opt.AllVersions:
-		return s.listAllVersions(ctx, key, prefix, limit, options)
-	default:
-		return s.listLastVersion(ctx, key, prefix, limit, options)
-	}
-}
+	// limit the number of results
+	opts = append(opts, clientv3.WithLimit(int64(opt.GetLimit())))
 
-// listPathsOnly returns only the path for each ruleset.
-func (s *RulesetService) listPathsOnly(ctx context.Context, key, prefix string, limit int, opts []clientv3.OpOption) (*api.Rulesets, error) {
-	opts = append(opts, clientv3.WithKeysOnly())
-	resp, err := s.Client.KV.Get(ctx, s.latestVersionPath(key), opts...)
+	// fetch signatures
```

Comment:
At first glance, it is not clear why you retrieve the signatures whereas the in-line doc says that only paths will be returned. I suggest to add more information here.
```suggestion
	// fetch signatures, the paths will be computed from signature keys
```